### DISCUSSION TITLE
Is there a WiFi bug with the PicoW w/ the 20241025 build?

### DISCUSSION THREAD

**Discussion by GM-Script-Writer-62850** [micropython/micropython#16288] [General]
I have a fairy simple script running on this PICO, i just switched it over to aiohttp from a modded urequest and updated to the 20241025 release and now i am getting error no 103 connection aborted when i make a http request to my local server (in 0-6 hours), last night this happened before 3AM, my server logs show it detected it was offline at 4AM and the pico checks every 5 min, in spite of it reconnecting the same error was still happening, before i make a aiohttp request i do check [isconnected()](https://docs.micropython.org/en/latest/library/network.html#network.AbstractNIC.isconnected)

if someone wants to see the source code: [Archive.zip](https://github.com/user-attachments/files/17872523/Archive.zip) (there is a bug with the buzzer activation when called from the http server, but this code was never executed)
this is where i am getting the error, this is part of a fn starting on line 35, wifi connection is managed by the fn starting on line 68

every time i have seen it go down it was running the main loop (fn is line 293, loop starts on line 314) aside from that all other stuff it idle waiting on input, this is the only thing trying to happen on the wifi chip

```
		async with aiohttp.ClientSession() as session:
			debug.net.connections+=1
			if data:
				async with session.post(page,data=json_encode(data),headers={"Content-Type":"application/json"}) as r:
					status=r.status
					content=await r.text()
			else:
				async with session.get(page) as r:
					status=r.status
					content=await r.text()
```

Note that i have another picoW doing way more complex stuff on the 20240602 release and it has been fine using aiohttp for months now

EDIT:
here is a screenshot from thonny: https://i.imgur.com/xdiVQ5j.png
note that nothing happened in my logs on my access point

**Comment by peterhinch** [comment]
You might be interested in [these observations](https://github.com/peterhinch/micropython-samples/blob/master/resilient/README.md) on maintaining a long term WiFi socket connection.

**Reply by GM-Script-Writer-62850** [reply]
the issue is on the older firmware it works till there is a power outage, had proven itself to be rock solid stable, the only time i had a issue was before i checked isconnected. I'm sure part of this is ubiquity APs are more reliable than a 40$ router from walmart, that said i never had a issue using one with my old second hand WZR300-HP (well until the wifi on it started crashing after ~7 years of service, now it is on switch duty)

if i am going from a mysterious issue once or twice a year (debugging that takes ages) to 4+ times a day i think there is a issue

**Reply by scruss** [reply]
But `isconnected()` will happily report True when the AP has gone away. It's not a reliable indicator of the link state.

The one detail that has changed recently is that you should turn off your MicroPython board's AP interface if you're using the station interface, and likely vice-versa.

Have you, or any of your neighbours, have a new or changed:
* wifi router?
* phone, tablet or computer?
* switch-mode power supply (can even be something as lowly as a wall-wart)?
* LED lighting? 
* microwave oven?

Any of these can mess with a wifi connection from quite a distance. Operational reliability in the [ISM](https://en.wikipedia.org/wiki/ISM_radio_band) bands is always somewhere between “maybe” and “`¯\_(ツ)_/¯`”.

**Reply by GM-Script-Writer-62850** [reply]
> The one detail that has changed recently is that you should turn off your MicroPython board's AP interface if you're using the station interface, and likely vice-versa.

if you are referring to using a PCIO as a AP i am not, it is a client device, if you are not talking about that idk what this is referring to

The only changes are the fw version and using aiohttp over urequest and reverting the fw made the issue go away

I switched the firmware back to the June release and it has worked ever since (aside from some glitch that happened seconds after the 1st power on where it detected it was offline and reconnected but was not working somehow) chopped that upto i unplugged and and reconnected power too fast to properly reset it (also changed a few lines in-case the recover issue was my fault, but that was detected)

i switched to the oct. firmware and i suddenly i get at best around 6 hours uptime, i go back to June firmware and the problem is gone

nothing has changed and it was going down with the house at full idle power usage, the only thing that could have turned on was the fridge that has been there since before the PICO was invented

also my cinder block wall as not kind to wireless (a blessing and a curse), and the distance is VERY short to my AP (locked to this AP BTW)
![20241124_124845](https://github.com/user-attachments/assets/4242f848-cf31-4258-82c6-fa6e316f82c8)

**Reply by scruss** [reply]
> if you are referring to using a PCIO as a AP i am not, it is a client device, if you are not talking about that idk what this is referring to

Do this, or something like it:

```py3
    sta_if = network.WLAN(network.STA_IF)
    ap_if = network.WLAN(network.AP_IF)
    if ap_if.active():
        ap_if.active(False)
        print("AP turned off ...")
```
Is there any wiring/metalwork under the Pico W's antenna? You want to keep that end part of the board clear of anything for best performance

**Reply by GM-Script-Writer-62850** [reply]
so `network.WLAN(network.AP_IF).active()===True` in recent builds when it was `False`?

this is how i have been doing wifi on my pico w units:

```
	from wifi_auth import ssid, password
	import network

	wlan = network.WLAN(network.STA_IF)
	wlan.active(True)
	wlan.connect(ssid,password)
	wlan.config(pm = 0xa11140)# Power management is very very bad, ping time is around 1000x worse and packet loss insane
```

**Reply by scruss** [reply]
Maybe best to make sure that the AP is turned off rather than checking.

The pm setting may have changed or the need for it gone away, too.

**Reply by GM-Script-Writer-62850** [reply]
this returns `False` on both builds:
`print(network.WLAN(network.AP_IF).active())`

as for the pm setting these results speak for themselves...
```
user@X470-Gaming-Plus:~$ ping -c 10 10.0.0.190
PING 10.0.0.190 (10.0.0.190) 56(84) bytes of data.
64 bytes from 10.0.0.190: icmp_seq=1 ttl=255 time=82.3 ms
64 bytes from 10.0.0.190: icmp_seq=2 ttl=255 time=1.72 ms
64 bytes from 10.0.0.190: icmp_seq=3 ttl=255 time=1.67 ms
64 bytes from 10.0.0.190: icmp_seq=4 ttl=255 time=1.72 ms
64 bytes from 10.0.0.190: icmp_seq=5 ttl=255 time=1.68 ms
64 bytes from 10.0.0.190: icmp_seq=6 ttl=255 time=1.70 ms
64 bytes from 10.0.0.190: icmp_seq=7 ttl=255 time=3.35 ms
64 bytes from 10.0.0.190: icmp_seq=8 ttl=255 time=1.69 ms
64 bytes from 10.0.0.190: icmp_seq=9 ttl=255 time=1.68 ms
64 bytes from 10.0.0.190: icmp_seq=10 ttl=255 time=4.72 ms

--- 10.0.0.190 ping statistics ---
10 packets transmitted, 10 received, 0% packet loss, time 9012ms
rtt min/avg/max/mdev = 1.671/10.225/82.320/24.051 ms
user@X470-Gaming-Plus:~$ ping -c 10 10.0.0.190
PING 10.0.0.190 (10.0.0.190) 56(84) bytes of data.
64 bytes from 10.0.0.190: icmp_seq=1 ttl=255 time=33.9 ms
64 bytes from 10.0.0.190: icmp_seq=2 ttl=255 time=46.1 ms
64 bytes from 10.0.0.190: icmp_seq=3 ttl=255 time=69.0 ms
64 bytes from 10.0.0.190: icmp_seq=4 ttl=255 time=92.3 ms
64 bytes from 10.0.0.190: icmp_seq=5 ttl=255 time=15.9 ms
64 bytes from 10.0.0.190: icmp_seq=6 ttl=255 time=34.8 ms
64 bytes from 10.0.0.190: icmp_seq=7 ttl=255 time=56.9 ms
64 bytes from 10.0.0.190: icmp_seq=8 ttl=255 time=79.0 ms
64 bytes from 10.0.0.190: icmp_seq=9 ttl=255 time=108 ms
64 bytes from 10.0.0.190: icmp_seq=10 ttl=255 time=23.7 ms

--- 10.0.0.190 ping statistics ---
10 packets transmitted, 10 received, 0% packet loss, time 9011ms
rtt min/avg/max/mdev = 15.921/55.990/108.288/29.039 ms

```

that said i am not dropping packet, so it is better than when i got my 1st pico w
